core/src/main/java/hudson/model/Queue.java

```diff
@@ -2443,29 +2485,45 @@ protected NotWaitingItem(NotWaitingItem ni) {
      * {@link Item} in the {@link Queue#blockedProjects} stage.
      */
     public final class BlockedItem extends NotWaitingItem {
+        private transient CauseOfBlockage causeOfBlockage = null;
+
+        /**
+         * @deprecated Use {@link #BlockedItem(WaitingItem, CauseOfBlockage)} instead
+         */
+        @Deprecated
         public BlockedItem(WaitingItem wi) {
-            super(wi);
+            this(wi, null);
         }
 
+        /**
+         * @deprecated Use {@link #BlockedItem(NotWaitingItem, CauseOfBlockage)} instead
+         */
+        @Deprecated
         public BlockedItem(NotWaitingItem ni) {
+            this(ni, null);
+        }
+
+        public BlockedItem(WaitingItem wi, CauseOfBlockage causeOfBlockage) {
+            super(wi);
+            this.causeOfBlockage = causeOfBlockage;
+        }
+
+        public BlockedItem(NotWaitingItem ni, CauseOfBlockage causeOfBlockage) {
             super(ni);
+            this.causeOfBlockage = causeOfBlockage;
         }
 
-        public CauseOfBlockage getCauseOfBlockage() {
-            ResourceActivity r = getBlockingActivity(task);
-            if (r != null) {
-                if (r == task) // blocked by itself, meaning another build is in progress
-                    return CauseOfBlockage.fromMessage(Messages._Queue_InProgress());
-                return CauseOfBlockage.fromMessage(Messages._Queue_BlockedBy(r.getDisplayName()));
-            }
+        public void setCauseOfBlockage(CauseOfBlockage causeOfBlockage) {
```

Comment:
Ditto. Why would this need to be a `public` API? What plugin would be calling it and why?